src/System.Private.CoreLib/shared/System/Type.Helpers.cs

```diff
@@ -129,27 +129,27 @@ public virtual Type[] FindInterfaces(TypeFilter filter, object filterCriteria)
                     cnt++;
             }
             if (cnt == c.Length)
-                return c;
+                return c!;
 
             Type[] ret = new Type[cnt];
             cnt = 0;
             for (int i = 0; i < c.Length; i++)
             {
                 if (c[i] != null)
-                    ret[cnt++] = c[i];
+                    ret[cnt++] = c[i]!;
```

Comment:
Another case where I don't understand why the `!` is necessary.  We just validated in the guarding `if` that `c[i]` isn't null, and now we have to use `c[i]!`?
cc: @jaredpar

Reply:
yeah it is from similar reason. Because of null assignment on row 127 ` c[i] = null;` was need to declare values of array c nullable `Type?[] c` which not allowing me to assign `c[i]` into non nullable `ret[cnt]`

Reply:
That shouldn't matter.  The code just validated that `c[i]` isn't null; we should not need to then use `c[i]!` then.  Seems like a gap in the compiler's analysis.

Reply:
This is a gap in the flow analysis although it is currently by design. The compiler tracks nullability of locals and parameters, and fields and properties of those variables, but not indexers.

---
In reply to: [270492785](https://github.com/dotnet/coreclr/pull/23489#discussion_r270492785) [](ancestors = 270492785)